### PR TITLE
Introduce reporting extensions

### PR DESCRIPTION
**Pull request by arturbosch**
This PR introduces the concept of a `ReportingExtension`.
With this extension the user can listen to results and manipulate them.

The first two implementations are `BaselineFacade` and `OutputFacade`.
With this concept the baseline and the reporting features do not need special treatment in the `Runner` class or the Intellij or Sonarqube plugin but are just loaded from the detekt core engine and executed.

No more code like:
```kotlin
result = arguments.baseline?.let { BaselineFacade().transformResult(it, result) } ?: result	                
val (outputResultsTime) = measure { OutputFacade(arguments.reportPaths, result, settings).run() }
```

Extracting the "special @Suppress" implemention to a `ReportingExtension` is conceivable in the future.
 
This PR also introduces a concept to share data between extensions and tooling frondends (cli, gradle, sonar, intellij) via the `PropertiesAware` interface.
It allows to retrieve data in a type safe way:
```kotlin
inline fun <reified T : Any> PropertiesAware.getOrNull(key: String): T?
```
This lightweight key-value storage reduces the need for extra parameters for the `ProcessingSettings` class.

I've marked both new api interfaces as `@UnstableApi` for now.

#2680

### CODE REVIEW

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/baseline/BaselineResultMapping.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import io.gitlab.arturbosch.detekt.api.SetupContext
+import io.gitlab.arturbosch.detekt.api.UnstableApi
+import io.gitlab.arturbosch.detekt.api.getOrNull
+import io.gitlab.arturbosch.detekt.core.DetektResult
+import java.nio.file.Path
+
+@OptIn(UnstableApi::class)
+class BaselineResultMapping : ReportingExtension {
+
+    private var baselineFile: Path? = null
```
Comment (BraisGabin) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/baseline/BaselineResultMapping.kt:15]: lateinit?

File: detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/runners/SingleRuleRunner.kt [hunk 2]
```diff
@@ -18,43 +16,31 @@ import java.io.PrintStream
...
-                listOf(DetektProgressListener().apply { init(settings) })
-            ).run()
-        }
+                DetektFacade.create(
+                    settings,
+                    listOf(provider),
+                    listOf(DetektProgressListener().apply { init(settings) })
+                ).run()
+            }
     }
```
Comment (BraisGabin) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/runners/SingleRuleRunner.kt:44]: Is this correctly fomatted?

File: detekt-cli/src/test/kotlin/io/gitlab/arturbosch/detekt/cli/runners/SingleRuleRunnerSpec.kt [hunk 3]
```diff
@@ -26,7 +24,7 @@ class SingleRuleRunnerSpec : Spek({
                 "--run-rule", "test:test"
             )
 
-            SingleRuleRunner(args, NullPrintStream(), NullPrintStream()).execute()
+            SingleRuleRunner(args, System.out, System.err).execute()
```
Comment (BraisGabin) [inline detekt-cli/src/test/kotlin/io/gitlab/arturbosch/detekt/cli/runners/SingleRuleRunnerSpec.kt]: This creates noisy test, right?

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/reporting/OutputFacade.kt [hunk 4]
```diff
@@ -1,36 +1,54 @@
...
-            .load()
-            .forEach(::handleOutputReport)
-        ConsoleReportLocator(settings)
-            .load()
-            .forEach(::handleConsoleReport)
+    override fun init(context: SetupContext) {
+        val reportPaths: Collection<ReportPath> =
+            context.getOrNull(DETEKT_OUTPUT_REPORT_PATHS_KEY) ?: emptyList()
+        reports = reportPaths.associateBy { it.kind }
+        settings = context as? ProcessingSettings ?: error("ProcessingSettings expected.")
```
Comment (BraisGabin) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/reporting/OutputFacade.kt:25]: 🤔 this cast is a bit strange, isn't it?

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/reporting/OutputFacade.kt [hunk 5]
```diff
@@ -13,6 +13,8 @@ import kotlin.system.measureTimeMillis
 @OptIn(UnstableApi::class)
 class OutputFacade : ReportingExtension {
 
+    override val priority: Int = Int.MIN_VALUE
```
Comment (BraisGabin) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/reporting/OutputFacade.kt:16]: Why is this necessary? OutputFacade only overrides `onFinalResult` and in that state no one should change the `Detektion` (for that reason I was talking about `Detektion` as an immutable class)

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/ReportingExtension.kt [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+    /**
+     * Allows to transform the reported findings e.g. apply custom filtering.
+     */
+    fun transformFindings(findings: Map<RuleSetId, List<Finding>>): Map<RuleSetId, List<Finding>> = findings
+
+    /**
+     * Is called after all extensions's [transformFindings] were called.
+     */
+    fun onFinalResult(result: Detektion) {
```
[the remark this answers is not in the document]
Comment (arturbosch) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/ReportingExtension.kt:25]: Yes, please feel free to write it down in the issue.

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/baseline/BaselineResultMapping.kt (hunk 1 above)
Comment (arturbosch) [reply detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/baseline/BaselineResultMapping.kt:15]: The baseline file can be null.

File: detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/runners/SingleRuleRunner.kt (hunk 2 above)
Comment (arturbosch) [reply detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/runners/SingleRuleRunner.kt:44]: Yes, still I extracted the long lambda to an function to make it more clear :).

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/reporting/OutputFacade.kt (hunk 5 above)
Comment (arturbosch) [reply detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/reporting/OutputFacade.kt:16]: However `Detektion` is still mutable. So it might be that anyone may add additional `ProjectMetric`'s and expect them to be printed on the console.

With running the output facade last we will ensure that everything stored will be printed.

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/reporting/OutputFacade.kt (hunk 4 above)
Comment (arturbosch) [reply detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/reporting/OutputFacade.kt:25]: Yes! I'm not sure how to handle this. It needs the plugin classloader. An extension which loads other extensions.
Feels kinda wrong.
I will convert it to a plain core class again.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/PropertiesAware.kt [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+ * Allows to retrieve stored properties in a type safe way.
+ */
+@UnstableApi
+inline fun <reified T : Any> PropertiesAware.getOrNull(key: String): T? {
+    val value = properties[key]
+    if (value != null) {
+        return value.safeAs() ?: error("No value of type ''${T::class} for key '$key'.")
+    }
+    return null
+}
```
[the remark this answers is not in the document]
Comment (arturbosch) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/PropertiesAware.kt:32]: Changed.
